lib/expr.mli

```diff
@@ -1,10 +1,11 @@
 (** Term definitions of the abstract syntax *)
-
-type t = private
+type inner = private
   { e : expr
   ; ty : Ty.t
   }
 
+and t = inner Hc.hash_consed
```

Comment:
Shouldn't we mark the three types here as `private` and only provide smart constructor ?

Reply:
I think expr can be left non private because it doesn't have any invariant other than those of its subtypes.

I'm not sure I get what you mean with `smart constructors`, currently `@:` does the trick when needing to create a new hasconsed value?

Reply:
I think `t` doesn't need to be `private` because `hash_consed` is, right? As for `expr`, for now, we can leave it as is. When I refactor the types per Arthur's recommendation, I can complete the `smart constructors` at the end of the file :smiley:

Reply:
Btw, do you think `Value.t` should be hash-consed as well?

Reply:
I discussed this with @lthls and it is indeed hash_consed that should be made private @zapashcanon

Reply:
> Btw, do you think `Value.t` should be hash-consed as well?

Only if a) they are used outside on an expr.Val *and* b) the string variant is used often with identical string I think.

Reply:
> Only if a) they are used outside on an expr.Val _and_ b) the string variant is used often with identicals string I think.

Oh okok, It's they are only ever used for models outside of `expr` and these values don't typically live for very long so it should be fine. I'm fine with merging this :)

Reply:
OK, it was not private in `hc` actually haha (I wrote it a long time ago :S). It should be fixed in https://git.zapashcanon.fr/zapashcanon/hc/commit/9a39243a3d010adca94a37abb2567bef749b1c07 I'll release it soon it the change is OK for you.

Reply:
> OK, it was not private in `hc` actually haha (I wrote it a long time ago :S). It should be fixed in https://git.zapashcanon.fr/zapashcanon/hc/commit/9a39243a3d010adca94a37abb2567bef749b1c07 I'll release it soon it the change is OK for you.

I don't have any objections :smiley: It looks great to me :)

Reply:
Should be good once https://github.com/ocaml/opam-repository/pull/25125 is merged.